www/js/lib/zimfile.js

```diff
@@ -238,6 +274,9 @@ define(['xzdec_wrapper', 'util', 'utf8', 'q', 'zimDirEntry'], function(xz, util,
                 zf.mimeListPos = readInt(header, 56, 8);
                 zf.mainPage = readInt(header, 64, 4);
                 zf.layoutPage = readInt(header, 68, 4);
+                zf._mimeTypeMap().then(function(data) {
+                    zf.mimeTypes = data;
```

Comment:
It would be safer to return the zf instance (through the Promise) after the `mimeTypes` map has been populated.
Else we get a race condition, where the zf could be read before we have the `mimeTypes` map

Reply:
OK, I may need help with this. I struggled quite a lot to find a good place to put it so that the `mimeTypeMap` would only be read once and would persist, but so that it would automatically be voided when the archive is changed. Part of the problem is that we need `zf.mimeListPos` to be available and populated before `zf._mimeTypeMap()` can be called...

Reply:
@mossroy To do this I would need to duplicated the code that calculates `zf.mimeListPos` and `zf.urlPtrPos`, and I wouldn't be able to use `zf._readSlice`. These are needed to access the MIME type map.

In practice, the MIME type map is read and stored within a few ticks of the archive being loaded. We don't need any MIME type string until the dirEntry has been read, which is much, much later. In the meantime, the zf object is exported and available to other processes that (might) need it, which is actually an advantage. Nothing else attempts to read the MIME type map other than a dirEntry, and then only after the dirEntry has itself been read, a process that lies behind several later promises...

Does it seem OK? I have tested this code a lot!

Reply:
It's true that the risk is very very low in the current context.
But it's a race condition anyway, it's better to fix it because it might occur in a different context (later version of the app, or a unit test for example)

Could you try with the following piece of code : 
```
                return zf._readMimetypeMap().then(function(data) {
                    zf.mimeTypes = data;
                    return zf;
                });
```

Reply:
@mossroy I had already tried that, and various other combinations. It's not possible to return zf inside the second promise, because that just defers resolving zf, but JS execution still continues of course, and `ZIMArchive.prototype.isReady` returns true before zf is returned. As a result, we get an immediate exception in `ZIMArchive.prototype.getMainPageDirEntry`, which thinks the ZIM Archive is ready, before we've actually constructed the file object!

I also tried chaining the promises this way:
```
            return util.readFileSlice(fileArray[0], 0, 80).then(function(header) {
                var mimeListPos = readInt(header, 56, 8);
                var urlPtrPos = readInt(header, 32, 8);
                readMimetypeMap(fileArray[0], mimeListPos, urlPtrPos).then(function(data) {
                    var zf = new ZIMFile(fileArray);
                    zf.articleCount = readInt(header, 24, 4);
                    zf.clusterCount = readInt(header, 28, 4);
                    zf.urlPtrPos = urlPtrPos;
                    zf.titlePtrPos = readInt(header, 40, 8);
                    zf.clusterPtrPos = readInt(header, 48, 8);
                    zf.mimeListPos = mimeListPos;
                    zf.mainPage = readInt(header, 64, 4);
                    zf.layoutPage = readInt(header, 68, 4);
                    zf.mimeTypes = data;
                    return zf;
                });
            });
```
However, `ZIMArchive.prototype.isReady` *still* returns true *before* the last promise above is resolved. In fact, doing it this way introduces a race condition. (The above code does actually produce the MIME type table, with a modification in `_readMimetypeMap` to use `util.readFileSlice` instead of `this.readSlice`.)

I'm afraid I've tried everything I can think of!

Reply:
PS to get the above code to run, I had to refactor `_readMimetypeMap` into a normal function (not part of the prototype) which is why it appears above referenced as `readMimetypeMap`. I can push a commit with that "working" (apart from the exception in `ZIMArchive.prototype.getMainPageDirEntry`) if you think you can fix it yourself.

My head is hurting from trying to understand the interaction between Objects and Promises here...

Reply:
@mossroy Please ignore!! I was missing a pesky return in the above code! I think I can push a working version now...

Reply:
Resolved!

Reply:
Cool, I did not test but it looks good